metomi/rose/config_editor/upgrade_controller.py

```diff
@@ -0,0 +1,329 @@
+        self.set_transient_for(parent_window)
+        self.window.set_title(metomi.rose.config_editor.DIALOG_TITLE_UPGRADE)
+        self.config_dict = {}
+        self.config_directory_dict = {}
+        self.config_manager_dict = {}
+        config_names = sorted(app_config_dict.keys())
+        self._config_version_model_dict = {}
+        self.use_all_versions = False
+        self.treemodel = Gtk.TreeStore(str, str, str, bool)
+        self.treeview = metomi.rose.gtk.util.TooltipTreeView(
+            get_tooltip_func=self._get_tooltip
+        )
+        self.treeview.show()
+        old_pwd = os.getcwd()
+        for config_name in config_names:
+            app_config = app_config_dict[config_name]["config"]
+            app_directory = app_config_dict[config_name]["directory"]
+            meta_value = app_config.get_value(
+                [
+                    metomi.rose.CONFIG_SECT_TOP,
+                    metomi.rose.CONFIG_OPT_META_TYPE,
+                ],
+                "",
+            )
+            if len(meta_value.split("/")) < 2:
+                continue
+            try:
+                os.chdir(app_directory)
+                manager = metomi.rose.upgrade.MacroUpgradeManager(app_config)
+            except OSError:
+                # This can occur when access is not allowed to metadata files.
+                continue
+            self.config_dict[config_name] = app_config
+            self.config_directory_dict[config_name] = app_directory
+            self.config_manager_dict[config_name] = manager
+            self._update_treemodel_data(config_name)
+        os.chdir(old_pwd)
+        self.treeview.set_model(self.treemodel)
+        self.treeview.set_rules_hint(True)
+        self.treewindow = Gtk.ScrolledWindow()
+        self.treewindow.show()
+        self.treewindow.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.NEVER)
+        columns = metomi.rose.config_editor.DIALOG_COLUMNS_UPGRADE
+        for i, title in enumerate(columns):
+            column = Gtk.TreeViewColumn()
+            column.set_title(title)
+            if self.treemodel.get_column_type(i) == GObject.TYPE_BOOLEAN:
+                cell = Gtk.CellRendererToggle()
+                cell.connect("toggled", self._handle_toggle_upgrade, i)
+                cell.set_property("activatable", True)
+            elif i == 2:
+                self._combo_cell = Gtk.CellRendererCombo()
+                self._combo_cell.set_property("has-entry", False)
+                self._combo_cell.set_property("editable", True)
+                self._combo_cell.connect(
+                    "changed", self._handle_change_version, 2
+                )
+                cell = self._combo_cell
+            else:
+                cell = Gtk.CellRendererText()
+            if i == len(columns) - 1:
+                column.pack_start(cell, True, True, 0)
+            else:
+                column.pack_start(cell, False, True, 0)
```

Comment:
I think that `pack_start` only takes 3 vars (including self.)

These extra variables arent in [the old code](https://github.com/metomi/rose/blob/2c8956a9464bd277c8eb24d38af4803cba4c1243/lib/python/rose/config_editor/upgrade_controller.py#L102C38-L103C53) and don't look like they match the [reference material](https://docs.gtk.org/gtk3/method.TreeViewColumn.pack_start.html)

Reply:
For long term reference, @astroDimitrios explained to me that `pack_start` can have 2 or 4 (non `self`) args depending on which widget the message is attached to. 😢